File: session4/lecture89/back_tracking.cpp
#include <bits/stdc++.h>
using namespace std;

int n, cnt, temp;
vector<int> v;
int ret;
const int mod = 11;

void go(int idx, int sum)
{
  // 벡트레킹 가지치기!!!
  if (ret == mod - 1)
    return;
  if (idx == n)
  {
    ret = max(ret, sum % mod);
    cnt++;
    return;
  }
  go(idx + 1, sum + v[idx]);
  go(idx + 1, sum);
}

int main()
{
  cin >> n;

  for (int i = 0; i < n; i++)
  {
    cin >> temp;
    v.push_back(temp);
  }

  go(0, 0);

  cout << ret << "\n";
  cout << cnt << "\n";

  return 0;
}

File: session4/lecture89/back_tracking.js
const input = require('fs').readFileSync('/dev/stdin', 'utf8').trim().split('\n');
let line = 0;
const readLine = () => input[line++];

const n = Number(readLine());
const v = readLine().split(' ').map(Number);

let ret = 0;
let cnt = 0;
const mod = 11;

function go(idx, sum) {
  // 벡트레킹 가지치기!!!
  if (ret === mod - 1) return;
  if (idx === n) {
    ret = Math.max(ret, sum % mod);
    cnt++;
    return;
  }
  go(idx + 1, sum + v[idx]);
  go(idx + 1, sum);
}

go(0, 0);

console.log(ret);
console.log(cnt);
